pkg/pesign/pesign.go

```diff
@@ -40,19 +40,82 @@ func NewSigner(provider CertificateSigner) (*Signer, error) {
 
 // Sign signs the input file and writes the output to the output file.
 func (s *Signer) Sign(input, output string) error {
+	if _, err := os.Stat(input); errors.Is(err, os.ErrNotExist) {
+		return fmt.Errorf("%s does not exist", input)
+	}
 	slog.Debug("Signing file", "input", input, "output", output)
-	unsigned, err := os.ReadFile(input)
+
+	si, err := os.Stat(input)
+	if err != nil {
+		return fmt.Errorf("failed getting input file info: %w", err)
+	}
+
+	ok, err := s.VerifyFile(input)
+	if ok {
+		slog.Warn("File is already signed with the cert, copying it into output file")
+		// already signed with the cert
+		// just copy it to the output place
+		f, err := os.ReadFile(input)
+		if err != nil {
+			return fmt.Errorf("failed reading input file: %w", err)
+		}
+		if err = os.WriteFile(output, f, si.Mode()); err != nil {
+			return fmt.Errorf("failed writing output file: %w", err)
+		}
+		return nil
+	}
+
+	peFile, err := os.Open(input)
+	if err != nil {
+		return err
+	}
+	defer peFile.Close()
+
+	peBinary, err := authenticode.Parse(peFile)
 	if err != nil {
-		slog.Error(fmt.Sprintf("Failed to open %s", input))
 		return err
 	}
-	signed, err := utils.SignEFIExecutable(s.provider.Signer(), s.provider.Certificate(), unsigned)
+
+	_, err = peBinary.Sign(s.provider.Signer(), s.provider.Certificate())
```

Comment:
the binary stuff from authertichode already provides a sign method for binaries :D